features/nvstore/source/nvstore.cpp

```diff
@@ -0,0 +1,1066 @@
+{
+}
+
+NVStore::~NVStore()
+{
+    if (_init_done) {
+        deinit();
+    }
+}
+
+uint16_t NVStore::get_max_keys() const
+{
+    return _max_keys;
+}
+
+void NVStore::set_max_keys(uint16_t num_keys)
+{
+    MBED_ASSERT(num_keys < MASTER_RECORD_KEY);
+    _max_keys = num_keys;
+    // User is allowed to change number of types. As this affects init, need to deinitialize now.
+    // Don't call init right away - it is lazily called by get/set functions if needed.
+    deinit();
+}
+
+// Flash access helper functions, using area and offset notations
+
+// Read from flash, given area and offset.
+// Parameters :
+// area          - [IN]   Flash area.
+// offset        - [IN]   Offset in area.
+// len_bytes     - [IN]   Length in bytes.
+// buf           - [IN]   Data buffer.
+// Return        : 0 on success. Error code otherwise.
+int NVStore::flash_read_area(uint8_t area, uint32_t offset, uint32_t len_bytes, uint32_t *buf)
+{
+    return nvstore_int_flash_read(len_bytes, _flash_area_params[area].address + offset, buf);
+}
+
+// Write to flash, given area and offset.
+// Parameters :
+// area          - [IN]   Flash area.
+// offset        - [IN]   Offset in area.
+// len_bytes     - [IN]   Length in bytes.
+// buf           - [IN]   Data buffer.
+// Return        : 0 on success. Error code otherwise.
+int NVStore::flash_write_area(uint8_t area, uint32_t offset, uint32_t len_bytes, const uint32_t *buf)
+{
+    return nvstore_int_flash_write(len_bytes, _flash_area_params[area].address + offset, buf);
+}
+
+// Erase a flash area, given area.
+// Parameters :
+// area          - [IN]   Flash area.
+// Return        : 0 on success. Error code otherwise.
+int NVStore::flash_erase_area(uint8_t area)
+{
+    return nvstore_int_flash_erase(_flash_area_params[area].address, _flash_area_params[area].size);
+}
+
+// Calculate area address and size configuration (in case it's not determined by the user)
+// or validate it (in case it is)
+void NVStore::calc_validate_area_params()
+{
+    int num_sectors = 0;
+
+    size_t flash_addr = nvstore_int_flash_get_flash_start();
+    size_t flash_size = nvstore_int_flash_get_flash_size();
+    size_t sector_size;
+    int max_sectors = flash_size / nvstore_int_flash_get_sector_size(flash_addr) + 1;
+    size_t *sector_map = new size_t[max_sectors];
+
+    int area = 0;
+    size_t left_size = flash_size;
+#if AREA_PARAMS_USER_CONFIG
+    _flash_area_params[0].address = NVSTORE_AREA_1_ADDRESS;
+    _flash_area_params[0].size    = NVSTORE_AREA_1_SIZE;
+    _flash_area_params[1].address = NVSTORE_AREA_2_ADDRESS;
+    _flash_area_params[1].size    = NVSTORE_AREA_2_SIZE;
+
+    int in_area = 0;
+    size_t area_size = 0;
+#endif
+    while (left_size) {
+        sector_size = nvstore_int_flash_get_sector_size(flash_addr);
+        //printf("Sector %3d: Address 0x%08x, size %6ld\n", num_sectors, flash_addr, sector_size);
+        sector_map[num_sectors++] = flash_addr;
+#if AREA_PARAMS_USER_CONFIG
```

Comment:
Why is this not validated the same way as a default location?

Reply:
Not sure I understand the comment. There are two options - either a user configuration for the area, which should be validated, or non-user configuration that needs to be calculated (and not validated of course).

Reply:
@davidsaada I do not fully understand neither `AREA_PARAMS_USER_CONFIG` what should do and how it should be configured (the macro does not have the name of the module it belongs to, what is params? parameters) ?

Reply:
@0xc0170 - We have two cases: 
1. (Default) No area address/size configuration was found in config files (e.g. mbed_app.json or mbed_lib.json). In this case, the code calculates address and size of each area.
2. Any of these parameters was found in config file by the user. In this case, the code validates the configuration. Defining any of these will set the define AREA_PARAMS_USER_CONFIG. If any of these is defined, code checks that ALL are defined.

Now, this define just tells the code whether it's case (1) or (2), that's all. Maybe its name is misleading, so would love to hear a suggestion for different name. Will add a comment there for clarification.